tests/compiletest.rs

```diff
@@ -1,71 +1,432 @@
+    path: &Path,
+    stderr: &str,
+    errors: &mut Errors,
+    require: bool,
+    revision: &str,
+) {
+    let content = std::fs::read_to_string(path).unwrap();
+    let mut found_annotation = false;
+    let regex =
+        Regex::new("//(\\[(?P<revision>[^\\]]+)\\])?~[\\^|]*\\s*(ERROR|HELP|WARN)?:?(?P<text>.*)")
+            .unwrap();
+    for line in content.lines() {
+        if let Some(s) = line.strip_prefix("// error-pattern:") {
+            let s = s.trim();
+            if !stderr.contains(s) {
+                errors.push(Error::PatternNotFound(s.to_string()));
+            }
+            found_annotation = true;
+        }
+        if let Some(captures) = regex.captures(line) {
+            // FIXME: check that the error happens on the marked line
+            let matched = captures["text"].trim();
+
+            if let Some(rev) = captures.name("revision") {
+                if rev.as_str() != revision {
+                    continue;
+                }
+            }
+
+            if !stderr.contains(matched) {
+                errors.push(Error::InlinePatternNotFound(matched.to_string()));
+            }
+            found_annotation = true;
+        }
+    }
+    if found_annotation != require {
+        errors.push(Error::NoPatternsFound);
+    }
+}
+
+fn check_output(
+    output: &[u8],
+    path: &Path,
+    errors: &mut Errors,
+    kind: String,
+    target: &str,
+) -> String {
+    let output = std::str::from_utf8(&output).unwrap();
+    let output = normalize(path, output);
+    let path = output_path(path, kind, target);
+    if env::var_os("MIRI_BLESS").is_some() {
+        if output.is_empty() {
+            let _ = std::fs::remove_file(path);
+        } else {
+            std::fs::write(path, &output).unwrap();
+        }
+    } else {
+        let expected_output = std::fs::read_to_string(&path).unwrap_or_default();
+        if env::var_os("MIRI_SKIP_UI_CHECKS").is_none() {
+            if output != expected_output {
+                errors.push(Error::OutputDiffers {
+                    path,
+                    actual: output.clone(),
+                    expected: expected_output,
+                });
+            }
+        }
+    }
+    output
+}
+
+fn output_path(path: &Path, kind: String, target: &str) -> PathBuf {
+    let content = std::fs::read_to_string(path).unwrap();
+    for line in content.lines() {
+        if line.starts_with("// stderr-per-bitwidth") {
+            return path.with_extension(format!("{}.{kind}", get_pointer_width(target)));
+        }
+    }
+    path.with_extension(kind)
+}
+
+fn compare_output(path: &Path, actual: &str, expected: &str) {
+    if actual == expected {
+        return;
+    }
+    eprintln!("actual output differed from expected {}", path.display());
+    eprintln!("{}", pretty_assertions::StrComparison::new(expected, actual));
+    eprintln!()
+}
+
+fn ignore_file(p: &Path, target: &str) -> bool {
+    let content = std::fs::read_to_string(p).unwrap();
```

Comment:
Wow how often do we read the same file into a string?^^

Reply:
Oh right, I was worried about this and it may in fact affect perf